react-js/react-components-and-hierarchies.md

```diff
@@ -0,0 +1,438 @@
+
+*Example*
+```
+class MyComponent extends React.Component {
+  render (){ 
+    return (
+      <div>
+        <h1>MyComponent JSX here</h1>
+        {this.props.children}
+      </div>
+    );
+  }
+}
+
+
+
+ReactDOM.render(
+  <MyComponent>
+    <p>understanding prop.children </p>
+  </MyComponent>,
+  document.getElementById('app')
+);
+```
+
+
+Whenever the MyComponent is invoked {props.children} will also be displayed and this is just a reference to what is between the opening and closing tags of the component.
+
+[Run code](https://codepen.io/annu12340/pen/wvKowmj?editors=1010)
+
+Instead of invoking the component with a self-closing tag < MyComponent /> if you invoke it will full opening and closing tags <MyComponent> </MyComponent> you can then place more code between it.
+ 
+This de-couples the <MyComponent> component from its content and makes it more reusable.
+
+The possible usage for {props.children} are:
+
+- Grouping unknown number of similar elements into a parent element.
+- You don’t know elements ahead of the time.
+- The nested structure that needs a wrapper.
+
+
+### Using 3p components like reactstrap
+1. Read [React Strap](https://reactstrap.github.io/) 
+2. Read [Add Reactstrap Components In ReactJS](https://www.c-sharpcorner.com/article/reactstrap-in-reactjs/)
+
+Reactstrap provides prebuilt Bootstrap 4 components that allow a great deal of flexibility and prebuilt validation. This allows us to quickly build beautiful forms that are guaranteed to impress and provide an intuitive user experience. It allows React developers to use various Bootstrap components such as grid system, navigation, icons, typography, forms, buttons, and table.
+
+#### Installation of reactstrap
+
+Reactstrap can be included directly in your application's bundle using a CDN
+```
+https://cdnjs.cloudflare.com/ajax/libs/reactstrap/4.8.0/reactstrap.min.js
+```
+Now, we are all set to use reactstrap UI components in React app.
+
+#### Using Reactstrap Buttons in React App
+Let’s check out how we can use reactstrap buttons in React app. First, we have to import Buttons component in src/App.js file and include the Buttons code from reactstrap site.
+
+```
+import React from 'react';
+import { Button } from 'reactstrap';
+import './App.css';
+
+class App extends React.Component{
+   render() {
+    return (
+    <div className="App">
+      <Button color="primary">primary</Button>
+      <Button color="secondary">secondary</Button>
+      <Button color="success">success</Button>
+      <Button color="info">info</Button>
+      <Button color="warning">warning</Button>
+      <Button color="danger">danger</Button>
+      <Button color="link">link</Button>
+    </div>
+  );
+   }
+
+}
+
+export default App;
+```
+[Run code](https://codepen.io/annu12340/pen/QWjpMRK?editors=1010)
+
+
+#### Reactstrap Cards
```

Comment:
This example is too verbose. Perhaps you can make it into an independent practice option?